File: python/src/toolnexus/http.py
"""HTTP / REST tools — declare a remote endpoint as a uniform Tool.

See ../../SPEC.md §7. Mirrors the JS reference (``js/src/http.ts``). Uses only the
standard library (``urllib``) so no new dependency is added; the blocking request
runs in a thread via ``asyncio.to_thread`` so it cooperates with the async toolkit.
"""
from __future__ import annotations

import asyncio
import json
import os
import re
import urllib.error
import urllib.parse
import urllib.request
from typing import Any, Optional

from .types import JSONSchema, Tool, ToolContext, ToolResult

DEFAULT_TIMEOUT = 30.0  # seconds (JS uses 30000 ms)

_EMPTY_SCHEMA: JSONSchema = {
    "type": "object",
    "properties": {},
    "additionalProperties": False,
}

_PLACEHOLDER_RE = re.compile(r"\{(\w+)\}")
_ENV_RE = re.compile(r"\$\{([A-Za-z0-9_]+)\}")


def _expand_headers(headers: Optional[dict[str, str]]) -> dict[str, str]:
    """Expand ${ENV_VAR} in header values from os.environ. Never logged."""
    out: dict[str, str] = {}
    for k, v in (headers or {}).items():
        out[k] = _ENV_RE.sub(lambda m: os.environ.get(m.group(1), ""), v)
    return out


def _safe_parse(text: str) -> Any:
    try:
        return json.loads(text)
    except Exception:
        return text


def http_tool(
    *,
    name: str,
    description: str,
    method: str,
    url: str,
    headers: Optional[dict[str, str]] = None,
    query: Optional[list[str]] = None,
    body: str = "json",  # "json" | "form" | "raw"
    input_schema: Optional[JSONSchema] = None,
    timeout: Optional[float] = None,
    result_mode: str = "text",  # "text" | "json" | "status+text"
) -> Tool:
    """Wrap an HTTP endpoint as a uniform :class:`Tool` (``source="http"``)."""
    upper_method = method.upper()
    query_set = set(query or [])
    schema = input_schema if input_schema is not None else _EMPTY_SCHEMA

    def _do_request(args: dict[str, Any], req_timeout: float) -> ToolResult:
        a = dict(args)

        # 1. substitute {placeholders} in the URL from args (consumed afterwards).
        consumed: set[str] = set()

        def _sub(m: "re.Match[str]") -> str:
            key = m.group(1)
            consumed.add(key)
            val = a.get(key)
            return urllib.parse.quote(str(val if val is not None else ""), safe="")

        final_url = _PLACEHOLDER_RE.sub(_sub, url)
        for key in consumed:
            a.pop(key, None)

        # 2. querystring args (named in `query`, or everything for GET/HEAD).
        params: list[tuple[str, str]] = []
        for key in list(a.keys()):
            if key in query_set or upper_method == "GET":
                params.append((key, str(a[key])))
                a.pop(key, None)
        if params:
            qs = urllib.parse.urlencode(params)
            final_url += ("&" if "?" in final_url else "?") + qs

        # 3. body
        req_headers = _expand_headers(headers)
        data: Optional[bytes] = None
        if upper_method not in ("GET", "HEAD") and a:
            if body == "json":
                req_headers.setdefault("Content-Type", "application/json")
                data = json.dumps(a).encode("utf-8")
            elif body == "form":
                req_headers.setdefault(
                    "Content-Type", "application/x-www-form-urlencoded"
                )
                data = urllib.parse.urlencode(a).encode("utf-8")
            else:  # raw
                data = str(a.get("body", "")).encode("utf-8")

        req = urllib.request.Request(
            final_url, data=data, method=upper_method, headers=req_headers
        )
        try:
            with urllib.request.urlopen(req, timeout=req_timeout) as resp:
                status = resp.status
                text = resp.read().decode("utf-8", errors="replace")
        except urllib.error.HTTPError as e:
            status = e.code
            text = e.read().decode("utf-8", errors="replace")
            return ToolResult(
                output=f"HTTP {status}: {text}",
                is_error=True,
                metadata={"status": status},
            )
        except Exception as e:  # noqa: BLE001 — surfaced as a tool error
            return ToolResult(output=str(e), is_error=True)

        if not (200 <= status < 300):
            return ToolResult(
                output=f"HTTP {status}: {text}",
                is_error=True,
                metadata={"status": status},
            )

        if result_mode == "status+text":
            output = f"{status}\n{text}"
        elif result_mode == "json":
            output = json.dumps(_safe_parse(text))
        else:
            output = text
        return ToolResult(output=output, is_error=False, metadata={"status": status})

    async def execute(
        args: Optional[dict[str, Any]] = None,
        ctx: Optional[ToolContext] = None,
    ) -> ToolResult:
        req_timeout = (
            ctx.timeout if (ctx is not None and ctx.timeout is not None) else timeout
        )
        if req_timeout is None:
            req_timeout = DEFAULT_TIMEOUT
        return await asyncio.to_thread(_do_request, args or {}, req_timeout)

    return Tool(
        name=name,
        description=description,
        input_schema=schema,
        source="http",  # type: ignore[arg-type]
        execute=execute,
    )

File: elixir/lib/toolnexus/http.ex
defmodule Toolnexus.Http do
  @moduledoc """
  HTTP / REST tools — declare a remote endpoint as a uniform `Toolnexus.Tool` (SPEC §7).

  Behavior mirrors the JS reference (`js/src/http.ts`):
    1. `{placeholder}`s in the URL are substituted from args (URL-encoded, consumed)
    2. args named in `:query` — or all remaining args on GET — become querystring params
    3. remaining args become the body (json | form | raw) on non-GET/HEAD requests
    4. `${ENV_VAR}` in header values expands from the environment at call time —
       expanded values are never logged
    5. non-2xx ⇒ `is_error` with `"HTTP <status>: <body>"`, otherwise the body text
  """

  alias Toolnexus.{Tool, ToolResult}

  @default_timeout 30_000
  @empty_schema %{"type" => "object", "properties" => %{}, "additionalProperties" => false}
  @placeholder ~r/\{(\w+)\}/
  @env_var ~r/\$\{([A-Z0-9_]+)\}/i

  @doc """
  Wrap an HTTP endpoint as a Tool.

  Options (keyword list or map):
    * `:name`, `:description`, `:method`, `:url` (required)
    * `:headers` — map of header name → value; values may contain `${ENV_VAR}`
    * `:query` — list of arg names sent as querystring instead of the body
    * `:body` — body encoding for non-GET requests: `"json"` (default), `"form"`, `"raw"`
    * `:input_schema` — JSON schema map (default: empty object schema)
    * `:timeout` — request timeout in ms (default 30000)
    * `:result_mode` — `"text"` (default), `"json"`, or `"status+text"`
    * `:req_options` — internal: extra `Req.new/1` options (e.g. `plug:` for tests)
  """
  @spec tool(keyword() | map()) :: Tool.t()
  def tool(opts) do
    opts = Map.new(opts)
    method = opts |> Map.fetch!(:method) |> to_string() |> String.upcase()

    cfg = %{
      method: method,
      url: Map.fetch!(opts, :url),
      headers: Map.get(opts, :headers) || %{},
      query_set: MapSet.new(Map.get(opts, :query) || []),
      body_mode: to_string(Map.get(opts, :body) || "json"),
      result_mode: to_string(Map.get(opts, :result_mode) || "text"),
      timeout: Map.get(opts, :timeout) || @default_timeout,
      req_options: Map.get(opts, :req_options) || []
    }

    %Tool{
      name: Map.fetch!(opts, :name),
      description: Map.fetch!(opts, :description),
      input_schema: Map.get(opts, :input_schema) || @empty_schema,
      source: "http",
      execute: fn args, _ctx -> run(cfg, args || %{}) end
    }
  end

  defp run(cfg, args) do
    # 1. substitute {placeholders} in the URL from args (consumed afterwards)
    {url, rest} = substitute(cfg.url, args)

    # 2. querystring args
    {pairs, rest} = split_query(rest, cfg.query_set, cfg.method)
    qs = URI.encode_query(pairs)
    url = if qs == "", do: url, else: url <> if(String.contains?(url, "?"), do: "&", else: "?") <> qs

    # 3. body
    headers = expand_headers(cfg.headers)
    {headers, body} = build_body(cfg.method, cfg.body_mode, headers, rest)

    req =
      Req.new(
        [
          method: method_atom(cfg.method),
          url: url,
          headers: headers,
          body: body,
          receive_timeout: round(cfg.timeout),
          retry: false,
          decode_body: false
        ] ++ cfg.req_options
      )

    case Req.request(req) do
      {:ok, %Req.Response{status: status, body: rbody}} ->
        text = to_text(rbody)

        if status in 200..299 do
          output =
            case cfg.result_mode do
              "status+text" -> "#{status}\n#{text}"
              "json" -> Jason.encode!(safe_parse(text))
              _ -> text
            end

          %ToolResult{output: output, is_error: false, metadata: %{status: status}}
        else
          %ToolResult{output: "HTTP #{status}: #{text}", is_error: true, metadata: %{status: status}}
        end

      {:error, e} ->
        %ToolResult{output: Exception.message(e), is_error: true}
    end
  rescue
    e -> %ToolResult{output: Exception.message(e), is_error: true}
  end

  # -- URL placeholder substitution (sequential, each key consumed like JS) ----

  defp substitute(url, args) do
    case Regex.run(@placeholder, url) do
      nil ->
        {url, args}

      [whole, key] ->
        val = args |> Map.get(key) |> str() |> encode_uri_component()
        url = String.replace(url, whole, val, global: false)
        substitute(url, Map.delete(args, key))
    end
  end

  # encodeURIComponent parity: unreserved chars plus ! * ' ( )
  defp encode_uri_component(s) do
    URI.encode(s, fn c -> URI.char_unreserved?(c) or c in ~c"!*'()" end)
  end

  defp split_query(args, query_set, method) do
    args
    |> Map.keys()
    |> Enum.sort()
    |> Enum.reduce({[], args}, fn key, {pairs, rest} ->
      if MapSet.member?(query_set, key) or method == "GET" do
        {pairs ++ [{key, str(Map.get(rest, key))}], Map.delete(rest, key)}
      else
        {pairs, rest}
      end
    end)
  end

  defp build_body(method, _mode, headers, _rest) when method in ["GET", "HEAD"],
    do: {headers, nil}

  defp build_body(_method, _mode, headers, rest) when map_size(rest) == 0,
    do: {headers, nil}

  defp build_body(_method, "form", headers, rest) do
    headers = Map.put_new(headers, "Content-Type", "application/x-www-form-urlencoded")
    {headers, rest |> Enum.map(fn {k, v} -> {k, str(v)} end) |> Enum.sort() |> URI.encode_query()}
  end

  defp build_body(_method, "raw", headers, rest), do: {headers, str(Map.get(rest, "body"))}

  defp build_body(_method, _json, headers, rest) do
    headers = Map.put_new(headers, "Content-Type", "application/json")
    {headers, Jason.encode!(rest)}
  end

  @doc false
  # Expand ${ENV_VAR} in header values from the environment (values never logged).
  def expand_headers(headers) do
    Map.new(headers, fn {k, v} ->
      {k, Regex.replace(@env_var, v, fn _, name -> System.get_env(name) || "" end)}
    end)
  end

  defp method_atom(method), do: method |> String.downcase() |> String.to_atom()

  defp safe_parse(text) do
    case Jason.decode(text) do
      {:ok, v} -> v
      {:error, _} -> text
    end
  end

  defp to_text(body) when is_binary(body), do: body
  defp to_text(body) when is_list(body), do: IO.iodata_to_binary(body)
  defp to_text(body), do: Jason.encode!(body)

  defp str(nil), do: ""
  defp str(v) when is_binary(v), do: v
  defp str(v), do: to_string(v)
end
